File: src/main/java/org/agentic/flink/corpus/Corpus.java
package org.agentic.flink.corpus;

import org.agentic.flink.context.core.ContextItem;
import org.agentic.flink.memory.vector.ScoredItem;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Named, possibly-shared knowledge base — vectors plus metadata, addressed by name.
 *
 * <p>A corpus is the framework abstraction for "an index of embedded {@link ContextItem}s that
 * multiple operators may read from and/or write to." Three flavours ship in-box:
 *
 * <ul>
 *   <li>{@link SingleOperatorCorpus} — both reads and writes happen in one operator. Smallest,
 *       fastest, simplest. The natural fit for the {@code KeyedCoProcessFunction} pattern.
 *   <li>{@link BroadcastCorpus} — ingest is one operator; reads happen in any number of read
 *       operators that each hold a replica via Flink broadcast state. Documented pattern; users
 *       wire the broadcast stream in their job graph.
 *   <li>{@link ExternalCorpus} — vectors live in a {@code VectorStore} (pgvector / Qdrant).
 *       Operators are stateless reads/writes against the shared external store.
 * </ul>
 *
 * <p>All three implement the same {@link Corpus} contract so a user can swap the flavour without
 * changing the agent code that consumes the corpus.
 */
public interface Corpus {

  /** Upsert a vectorized item. */
  CompletableFuture<Void> upsert(String id, float[] embedding, ContextItem item);

  /** Search the top-k most similar items to {@code query}. Higher score = more similar. */
  CompletableFuture<List<ScoredItem>> search(float[] query, int k);

  /** Snapshot of size + dimension + provider. */
  CorpusStats stats();

  /** Free runtime resources. Default: no-op. */
  default void close() throws Exception {
    // no-op
  }
}

File: python/agentic_flink/corpus.py
"""Corpus specs — three flavours, all returning live Java ``CorpusSpec`` instances."""

from __future__ import annotations

from typing import Mapping

from ._jvm import jclass


def single_operator(name: str, vector_spec):
    """In-operator corpus. Both reads and writes happen on the same Flink
    operator (typically a ``KeyedCoProcessFunction``)."""
    Single = jclass("org.agentic.flink.corpus.SingleOperatorCorpus")
    java_vec = vector_spec._to_java() if hasattr(vector_spec, "_to_java") else vector_spec
    return Single.spec(name, java_vec)


def broadcast(name: str, vector_spec):
    """Ingest in one operator; per-subtask replicas in any number of read
    operators via Flink broadcast state. User wires the broadcast plumbing;
    each replica is a :func:`single_operator` view underneath."""
    Broadcast = jclass("org.agentic.flink.corpus.BroadcastCorpus")
    java_vec = vector_spec._to_java() if hasattr(vector_spec, "_to_java") else vector_spec
    return Broadcast.spec(name, java_vec)


def external(
    name: str,
    backend: str,
    backend_config: Mapping[str, str],
    dimension: int,
):
    """External vector store (``pgvector``, ``qdrant``, …). Operators are
    stateless; the corpus is shared across jobs."""
    External = jclass("org.agentic.flink.corpus.ExternalCorpus")
    HashMap = jclass("java.util.HashMap")
    cfg = HashMap()
    for k, v in backend_config.items():
        cfg.put(str(k), str(v))
    return External.spec(name, backend, cfg, int(dimension))


__all__ = ["single_operator", "broadcast", "external"]
